### ISSUE TITLE
Wrong sorting of filenames in Ukrainian

### ISSUE THREAD

**Issue by mc-butler** [MidnightCommander/mc#4407]
> [!IMPORTANT]
> This issue was migrated from Trac:
>
> | **Origin** | https://midnight-commander.org/ticket/4407 |
> | :------------- | :------------------------------------------------------------------------------------------------ |
> | **Reporter** | `aj` (acid@….kiev.ua) |
> | **Keywords** | `sorting`, `order`, `Ukrainian` |

<p>
Hello.<br />
mc sorts filenames in Ukrainian language in wrong order.<br />
</p>
<p>
You can compare order with <a class="ext-link" href="https://en.wikipedia.org/wiki/Ukrainian_alphabet">https://en.wikipedia.org/wiki/Ukrainian_alphabet</a><br />
</p>
<p>
Package is from Debian and has version 3:4.8.26-1.1.<br />
</p>
<p>
$ LC_MESSAGES=C mc -V<br />
GNU Midnight Commander 4.8.26<br />
Built with GLib 2.66.8<br />
Built with S-Lang 2.3.2 with terminfo database<br />
With builtin Editor and Aspell support<br />
With subshell support as default<br />
With support for background operations<br />
With mouse support on xterm and Linux console<br />
With support for X11 events<br />
With internationalization support<br />
With multiple codepages support<br />
With ext2fs attributes support<br />
Virtual File Systems:<br />
</p>
<blockquote>
<p>
cpiofs, tarfs, sfs, extfs, ext2undelfs, ftpfs, sftpfs, fish<br />
</p>
</blockquote>
<p>
Data types:<br />
</p>
<blockquote>
<p>
char: 8; int: 32; long: 64; void *: 64; size_t: 64; off_t: 64;<br />
</p>
</blockquote>
<p>
$ LC_MESSAGES=C mc -F<br />
mc --configure-options<br />
Home directory: /home/jack<br />
Profile root directory: /home/jack<br />
</p>
<p>
[System data]<br />
</p>
<blockquote>
<p>
Config directory: /etc/mc/<br />
Data directory:   /usr/share/mc/<br />
File extension handlers: /usr/lib/mc/ext.d/<br />
VFS plugins and scripts: /usr/lib/mc/<br />
</p>
<blockquote>
<p>
extfs.d:        /usr/lib/mc/extfs.d/<br />
fish:           /usr/lib/mc/fish/<br />
</p>
</blockquote>
</blockquote>
<p>
[User data]<br />
</p>
<blockquote>
<p>
Config directory: /home/jack/.config/mc/<br />
Data directory:   /home/jack/.local/share/mc/<br />
</p>
<blockquote>
<p>
skins:          /home/jack/.local/share/mc/skins/<br />
extfs.d:        /home/jack/.local/share/mc/extfs.d/<br />
fish:           /home/jack/.local/share/mc/fish/<br />
mcedit macros:  /home/jack/.local/share/mc/mc.macros<br />
mcedit external macros: /home/jack/.local/share/mc/mcedit/macros.d/macro.*<br />
</p>
</blockquote>
<p>
Cache directory:  /home/jack/.cache/mc/<br />
</p>
</blockquote>
<blockquote>
<p>
'--build=x86_64-linux-gnu' '--prefix=/usr' '--includedir=${prefix}/include' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--sysconfdir=/etc' '--localstatedir=/var' '--disable-option-checking' '--libdir=${prefix}/lib/x86_64-linux-gnu' '--runstatedir=/run' '--disable-maintainer-mode' '--disable-dependency-tracking' 'AWK=awk' 'X11_WWW=x-www-browser' '--libexecdir=/usr/lib' '--with-x' '--with-screen=slang' '--disable-rpath' '--disable-static' '--disable-silent-rules' '--enable-aspell' '--enable-vfs-sftp' '--enable-vfs-undelfs' '--enable-tests' 'build_alias=x86_64-linux-gnu' 'CFLAGS=-g -O2 -ffile-prefix-map=/build/mc-xlv0ck/mc-4.8.26=. -fstack-protector-strong -Wformat -Werror=format-security' 'LDFLAGS=-Wl,-z,relro -Wl,-z,now -Wl,--as-needed' 'CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2'<br />
</p>
</blockquote>

> [!NOTE]
> Original attachments:
>
> * [Знімок екрану_2022-09-19_14-48-55.png](https://github.com/MidnightCommander/trac-archive/blob/master/attachments/ticket/4407/%D0%97%D0%BD%D1%96%D0%BC%D0%BE%D0%BA%20%D0%B5%D0%BA%D1%80%D0%B0%D0%BD%D1%83_2022-09-19_14-48-55.png) ([raw](https://github.com/MidnightCommander/trac-archive/raw/master/attachments/ticket/4407/%D0%97%D0%BD%D1%96%D0%BC%D0%BE%D0%BA%20%D0%B5%D0%BA%D1%80%D0%B0%D0%BD%D1%83_2022-09-19_14-48-55.png)) by `aj` (acid@….kiev.ua) on <span title="2022-09-19T11:54:32+00:00">`Sep 19, 2022 at 11:54 UTC`</span>

**Comment by egmontkob** [comment]
I'm facing similar problems with Hungarian sorting (`LC_ALL=hu_HU.UTF-8`).

Hungarian has quite complicated collation rules. Apart from an implementation thereof, glibc also contains an extensive unittest in its [localedata/hu_HU.UTF-8.in](https://github.com/bminor/glibc/blob/master/localedata/hu_HU.UTF-8.in) file. ((( Guess who made it 😛 )))

Creating files with those names, the sorting order in mc is different from what's in the file.

---

There's no `strcasecoll()` method, so case insensitive sorting definititely needs hacks: something like converting all the filenames to lowercase (or all to uppercase) and calling `strcoll()` on those.

~Case sensitive~ (edit) case insensitive sorting should just result in whatever `strcoll()` tells without any further hacks.

**Comment by egmontkob** [comment]
> There's no `strcasecoll()` method

I was wrong... semantically...

There's case-sensitive `strcmp()` and case-insensitive `strcasecmp()` that we're all familiar with. The first is strict byte ordering, the second makes it case insensitive via some more complex logic.

`strcoll()` is also case-insensitive, at least in the locales I've checked right now, and presumably in most (if not all) locales. No sane language would come up with dictionaries that first list all the capital-initial words and then all the lowercase ones, or the other way around. Case-insensitive sorting is the natural for humans, and that's what `strcoll()` aims to implement.

So, `strcmp()` has no locale-aware counterpart, whereas `strcasecmp()`'s locale-aware counterpart is `strcoll()`.

In libc, that is.

On top of this go a couple of GLib methods, in turn on top of which go a couple of mc methods...

**Comment by egmontkob** [comment]
The original report about Ukrainian, and my followup about Hungarian, are unrelated.

---

The Ukrainian issue is about _case sensitive_ order.

Case sensitive, locale-aware sorting is not offered by our underlying libraries libc and GLib, nor do we hack up something.

`src/filemanager/dir.c` `compare_by_names()` first caches (if not already cached) two comparison keys by calling `str_create_key_for_filename()`, which is in fact `lib/strutil/strutilutf8.c` `str_utf8_create_key()`. This, in case sensitive mode, turns accented letters to their NFD form, i.e. unaccented base letter followed by a combining accent. 

`compare_by_names()` then calls `key_collate()` which is just a simple wrapper to sort dot-files to the top, calling `str_key_collate()` which is in fact `str_utf8_key_collate()` which just calls the locale-unaware byte-by-byte `strcmp()`.

So, a letter like `ö` is decomposed to `o` + combining accent (not sure if some Ukrainian letters are also affected here, but this is not the point), and then these are compared as raw byte strings (resulting in Unicode character code order).

And it just so happens that Unicode order is the one shown in the screenshots, and is not the one that Ukrainian speakers are looking for.

We should try to figure out if a locale-aware case sensitive sorting could be implemented on top of the locale-aware case insensitive sorting that we have access to. Or not offer this feature, maybe reword the UI elements, maybe even change the default.

---

The Hungarian issue is about _case insensitive_ order. (Sure it has issues in _case sensitive_ mode as well, related to the previous story, but here I'm not talking about those.)

Lowercase and uppercase variants of the same word are sorted in the wrong order. The official sorting rules of the language describe which one goes first, and glibc carefully implements and tests this. In mc this happens to become the other order.

`str_utf8_create_key_gen()`, in _case insensitive_ mode, calls `g_utf8_casefold()` followed by `g_utf8_collate_key()` disguised as `keygen()` multiple times. Why it's multiple times rather than just once is irrelevant: it's to tolerate possible invalid UTF-8 fragments. What matters is that it should only be a `keygen()` a.k.a. `g_utf8_collate_key()` call and let it take care of lowercase vs. uppercase, without us explicitly `g_utf8_casefold()`ing first because that's where we break the ordering of lowercase vs. uppercase.

That is, if we verify that pretty much all collation rules implement case insensitive ordering. Because if some implement case sensitive then we're a bit in trouble: can we detect that, how would we communicate that on the UI? I mean, we couldn't label the backing algorithm as _case sensitive_ nor as _case insensitive_ as it'd depend on the locale.

---

For both issues:

_Maybe_ instead of calling the two options _case sensitive_ and _case insensitive_, we should call them _character code order_ and _language order_, or something similar.

Actually now we call it _case sensitive_ or lack thereof (checkbox), we might need to convert that to a two-state radiobutton to be able to denote what it means if the option is unset, because by seeing just one (either one) of _character code order_ or _language order_ it's probably not obvious what the other behavior is.

**Comment by egmontkob** [comment]
> We should try to figure out if a locale-aware case sensitive sorting could be implemented on top of the locale-aware case insensitive sorting that we have access to.

Here's my idea, if we want to try to do this:

Comparing two strings would go like:

Find the first position where they differ.

If here one of them has a lowercase letter while the other has an uppercase one, return the order accordingly.

(I'm quite certain that we'd also need to make a decision if one of them has an EOS or a non-letter character here. Otherwise, if we'd fall back to the locale-aware comparison in these cases, I'm fairly certain that there could be rock-paper-scissors loops.)

If, at the first difference, both strings have a lowercase letter or both strings have an uppercase letter, then continue with comparing the sorting key, i.e. whatever the locale-aware `strcoll()` / `g_utf8_collate()` would tell us.

Obviously we'd also need to safely handle invalid UTF-8 filenames.

Is it worth trying it? I don't think so. I guess just offering one single locale-aware sorting (varying based on the locale) should be fine.

**Comment by acid-jack** [comment]
Oh. You migrated to the github.
Problem with wrong sort order of Ukrainian letters is still present. But I found interesting behavior. If I disable 'Сase sensitive' option for 'Sort order' in the panel settings then sorting is correct!
And if this option is enabled then 'э', 'і', 'ї', 'ґ' goes to the end after 'я'  and 'Є', 'І', 'Ї' goes before 'А' and 'Ґ' goes after 'Я'.

So I don't understand how 'Sort order' option for panels should work.
